File: ruby/phone-number/phone_number.rb
class PhoneNumber
  attr_reader :number

  VALID_CHARACTERS = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']

  def initialize(number)
    @number = clean_nubmer(number)
  end

  def clean_nubmer(number)
    number.gsub!(/\W/, "")
    if number.chars.any? { |digit| !VALID_CHARACTERS.include?(digit) }
      '0000000000'
    elsif number.length == 11 && number.start_with?("1")
      number.chars[1..-1].join
    elsif number.length != 10
      '0000000000'
    else
      number
    end
  end

  def area_code
    number[0..2]
  end

  def to_s
    "(#{number[0..2]}) #{number[3..5]}-#{number[6..-1]}"
  end
end

File: javascript/phone-number/phone-number.js
function PhoneNumber(initial_number) {
  this.number = function() {
    var number = initial_number.replace(/\D/g, "");
    return this.validateNumber(number);
  }

  this.validateNumber = function(number) {
    if (number.length === 10) {
      return number
    } else if(number.length === 11 && number[0] == '1') {
      return number.slice(1, 11);
    } else if (number.length >= 11 || number.length <= 9) {
      return '0000000000';
    }
  }

  this.areaCode = function() {
    return this.number().substring(0, 3);
  }

  this.toString = function() {
    return '(' + this.areaCode() + ') ' + this.number().substring(3, 6) + '-' + this.number().substring(6, 11)
  }
}

module.exports = PhoneNumber
